### ISSUE TITLE
Add analyzer to find incorrect use of locks in async methods

### ISSUE THREAD

**Issue by mattwar** [dotnet/roslyn-analyzers#314]
Any use of a synchronous lock across an asynchronous await is a bug waiting to happen.

The following is badness:

```cs
private int protectedField;
private SemaphoreSlim mylock;

public async Task DoThatThing()
{
     using (mylock.DisposableWait())
     {
           protectedField = await MyOtherThing().ConfigureAwait(false);
     }
}
```

The correct way:

```cs
private int protectedField;
private SemaphoreSlim mylock;

public async Task DoThatThing()
{
     using (await mylock.DisposableWaitAsync().ConfigureAwait(false))
     {
           protectedField = await MyOtherThing().ConfigureAwait(false);
     }
}
```

**Comment by Evangelink** [comment]
@mavasani Is it ok to start working on this rule?

**Comment by Evangelink** [comment]
Probably needs to get through the `runtime` team check, cc @mavasani

**Comment by Youssef1313** [comment]
This was labeled `Area-Roslyn.Diagnostics.Analyzers`, but I don't think it's dotnet/roslyn specific.

@stephentoub Should this move to dotnet/runtime?
